Option to hide videos from certain channels

petaded opened the pull request:
# Option to hide videos from certain channels

<!-- Thanks for sending a pull request! Make sure to follow the contributing guidelines. -->
<!-- Important note, we may remove your pull request if you do not use this provided PR template correctly. -->

## Pull Request Type
<!-- Please select what type of pull request this is: [x] -->
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
<!-- Please link the issue your pull request is referring to. -->
<!-- If this pull request fully resolves the relevant issue, put "closes" before the issue number. -->
<!-- Example: "closes #123456". -->
partial implementation of #95

## Description
This PR adds a new setting under distraction free settings to hide specific channels.
You can write either the channel name or the channel id into the input box - multiple channels allowed via separating with semi-colon ;

Videos and playlists from this channel ( and the channel itself ) will no longer show up in a search, recommended or trending.
Note: the box is case sensitive and must be an exact match to the channel name.

## Screenshots <!-- If appropriate -->
<!-- Please add before and after screenshots if there is a visible change. -->
The new setting ( see bottom right ):
![image](https://user-images.githubusercontent.com/74797538/200912235-b0539166-c2b2-4c0a-8156-d2a1d2034512.png)
Option box with multiple channels entered
![image](https://user-images.githubusercontent.com/74797538/200913015-94670103-c441-4093-b4c1-651fab59553f.png)

Trending with empty box
![image](https://user-images.githubusercontent.com/74797538/200912356-13fcab75-0df7-471b-848c-bf31bb9a2f2e.png)
Trending with "shane" in hide channel box
![image](https://user-images.githubusercontent.com/74797538/200912441-754503ed-2ada-4b7b-b206-4d26fb0c9da1.png)
Tending after having "shane;A24" in hide channels box
![image](https://user-images.githubusercontent.com/74797538/200913126-ea5c9833-250c-47e1-93fc-267b29dbf809.png)



## Testing <!-- for code that is not small enough to be easily understandable -->
<!-- Has this pull request been tested? -->
<!-- Please describe shortly how you tested it. -->
<!-- Are there any ramifications remaining? -->
Tested the following:
 - since this touched the livestream filtering code
   - Search "music live" with hide livestreams as false, validate live streams still show
   - Search "music live" with hide livestreams as true, validate live streams no longer show
 - Search "ltt" with empty Hide channels, ltt videos show
 - Search "ltt" with "Linus Tech Tips" in hide channel box, ltt videos hidden + channel also hidden
 - Search "ltt" with "Linus Tech Tips;LMG Clips" in hide channel box, ltt and lmg clips videos hidden
 - Search "ltt" with "UCXuqSBlHAE6Xw-yeJA0Tunw" in hide channel box ( ltt channel id), ltt videos hidden



## Desktop
<!-- Please complete the following information-->
- **OS:** Manjaro 
- **OS Version:** 22
- **FreeTube version:** 0.18.0

## Additional context
<!-- Add any other context about the pull request here. -->

PikachuEXE commented:
Idea wise it's useful
But UI/UX wise it's inconsistent with other "distraction free" settings
(It might even deserve a dedicated section if we would have more similar settings in the future)

- There is no way I can tell it's about hiding channels
- FT has no previous input for multiple text values, the UI for such new input type requires some discussion to agree on
  But having a text box to have value separated by a separator is certainly unacceptable for me

I imagine the UI at least look like the following
![image](https://user-images.githubusercontent.com/1018543/201245156-e86c1062-4a29-4bc6-b331-1d5b53d1d195.png)

https://dribbble.com/shots/5970633-Add-Tags

It should handle many values (10+) and probably also mobile friendly
But let's make it at least work for desktop for the 1st version

This is only my opinion so would need more people to comment on this

petaded commented:
@PikachuEXE I would completely agree.

I posted it as it was because:
1. I happened to want this feature recently so coded something up for myself, was sharing this in case its useful for someone else
2. I have not done any javascript, electron, vue programming before, so was trying to limit the scope of my changes

However I think you are right, it shouldn't be included as is, the current design is not intuitive.
I am happy for anyone else to add to this PR otherwise I'll have a look into styling a new input type when I get some more time.

miangraham commented:
> It might even deserve a dedicated section if we would have more similar settings in the future

I've been thinking along similar lines as I'm considering adding a "never show me shorts" setting, which could just as easily be a "filter by title keywords" setting since YT just has the tag sitting in the title.

It would make sense for video-hiding settings to move to a new category rather than sitting in Distraction Free Settings. It's a consistent structural divide between options that toggle dedicated UI components versus options that filter items from lists. And it would do a bit to trim Distraction Free's massive wall of switches.

If my other PR lands soon I may try pushing this one for a bit. I'm pretty new to Vue but I imagine there are tag list component examples floating around.

petaded commented:
I was looking at this a bit today. have made a "ft-input-tags" component - have not yet done the logic was just sorting the look.
Will update the code when I have it working.

Currently looks like this :
![image](https://user-images.githubusercontent.com/74797538/201526330-67ebaede-89f9-47a2-b98c-4fbad0f67fb1.png)
![image](https://user-images.githubusercontent.com/74797538/201526343-46fcd6a8-5fd4-4906-bc5c-7b7014ea85d4.png)

```
<ft-input-tags
          :title="$t('Settings.Distraction Free Settings.Hide Channels Title')"
          :placeholder="$t('Settings.Distraction Free Settings.Hide Channels Placeholder')"
          :show-action-button="true"
          :value="hideSpecificChannels"
        />
```

miangraham commented:
Looks great. Thanks for posting to save the duplicate work. Will definitely want to use that in the future myself.

PikachuEXE commented:
@petaded Looks good so far!
Please let us know when done (enough for testing)

petaded commented:
Have updated PR. 
Added a new ft-input-tags input component. The new option now uses this component. The logic for hiding videos etc is still the same just the input method is updated.

Here is a short video of using it.
https://user-images.githubusercontent.com/74797538/201694537-b28978f5-085a-4aaa-832b-a70d7e3b2f61.mp4

If someone could test on windows that would be useful ( I don't have access to a windows computer )

I still have it under distraction free settings but only because I am not sure what/where the option should go, some more input on this would be welcome.

PikachuEXE commented:
The new UI element looks and works fine in simple test

Though I am not sure what value I should put into it (Channel name can change but I guess rarely?)
Does channel ID work? Does partial name work?
I can never tell until I test it or look at the code and this is bad for UX.

For the position I would say put it as a separate box below the columns of switches (so the width = all columns)

petaded commented:
@PikachuEXE  Thank you for the feedback.
I have added an optional tooltip parameter to the ft-input-tags component. Then added some tooltip text to the distraction free setting see picture below:
![image](https://user-images.githubusercontent.com/74797538/201902670-1fa488e5-ea8a-4089-abb0-cc7be1b9eab1.png)
I also updated the placeholder text to be "Channel Name or ID" to make it more obvious without reading the tooltip that you can also use a channel id.

Also as suggested have moved the setting to below the other toggle switches.
![image](https://user-images.githubusercontent.com/74797538/201903092-3efd4c6e-f57c-4d8b-a9a9-5159647da9e0.png)

ChunkyProgrammer commented:
It's not very accessible at the moment (labels should be linked to the inputs)

petaded commented:
Thanks @ChunkyProgrammer for having a look, however I don't quite understand what you mean by labels should be linked to inputs - I'm guessing you are talking about the code?

Do you have specific suggestions of what I could change to improve it?

PikachuEXE commented:
I agree on the semicolon separator issue
Need to use another "won't be used on channel name" char
But I have no idea what to use :)

petaded commented:
I've made the following updates:

- Changed from split/join on ';' to using `JSON.stringify` and `JSON.parse`
    - This results in any special characters being usable.
- added a `@click="focusInput"` to the label
   - I couldn't use for=ID since the input is inside a child component therefore cannot access its ID at html generation time, so I instead mimicked the functionality with a simple focus function
   - I had not heard of the label for attribute before, this is useful to know for the future thank you :)
 - changed the `title=XX` property to `label=XX` so its inline with the other components. e.g. ft-toggle-switch

Some symbol testing.
![image](https://user-images.githubusercontent.com/74797538/202166048-8c4dd9f4-11c2-445c-a837-d0c36fc119d8.png)
More outrageous symbol testing
![image](https://user-images.githubusercontent.com/74797538/202168521-11e196c6-8dcf-4e45-9bfb-7235c088bedf.png)

Focus testing
https://user-images.githubusercontent.com/74797538/202167699-0d46f087-22e5-4949-bc27-6e1a75ce5012.mp4

Note: if you have a value stored atm you will get an error due to the ';' to stringify change. Either clear your settings or temporarily add a `return []` to `hideSpecificChannels` in distraction-settings and add a random value, then remove this line.

PikachuEXE commented on src/renderer/components/ft-input-tags/ft-input-tags.css:
```diff
@@ -0,0 +1,62 @@
+  padding-left: 10px;
+  overflow-wrap: break-word;
+  word-wrap: break-word;
+  word-break: break-all;
+  word-break: break-word;
```
duplicate properties

petaded replied:
Have removed, nice spot!

PikachuEXE commented on src/renderer/store/modules/settings.js:
```diff
@@ -195,6 +195,7 @@ const state = {
   hideChannelSubscriptions: false,
   hideCommentLikes: false,
   hideComments: false,
+  hideSpecificChannels: '',
```
This is not a bool option, a channel list/collection is what it is
```suggestion
  ChannelsHidden: '',
```

PikachuEXE commented:
For the label thing I tried using ft-input component
Requires a bit modification to it to support custom label but better than `@click="focusInput"`

<details><summary>Patch</summary>

```patch
From 14137fd242123a3752f256799d811e90fd8f832c Mon Sep 17 00:00:00 2001
From: PikachuEXE <pikachuexe@gmail.com>
Date: Thu, 17 Nov 2022 08:53:02 +0800
Subject: [PATCH] test

---
 .../components/ft-input-tags/ft-input-tags.css  |  4 +++-
 .../components/ft-input-tags/ft-input-tags.js   |  3 ---
 .../components/ft-input-tags/ft-input-tags.vue  | 17 +++--------------
 src/renderer/components/ft-input/ft-input.js    |  4 ++++
 src/renderer/components/ft-input/ft-input.vue   |  2 +-
 5 files changed, 11 insertions(+), 19 deletions(-)

diff --git a/src/renderer/components/ft-input-tags/ft-input-tags.css b/src/renderer/components/ft-input-tags/ft-input-tags.css
index 9433cee6..475b6e69 100644
--- a/src/renderer/components/ft-input-tags/ft-input-tags.css
+++ b/src/renderer/components/ft-input-tags/ft-input-tags.css
@@ -59,4 +59,6 @@
   cursor: pointer;
 }
 
-
+:deep(.ft-input-component .ft-input) {
+  margin-top: 10px;
+}
diff --git a/src/renderer/components/ft-input-tags/ft-input-tags.js b/src/renderer/components/ft-input-tags/ft-input-tags.js
index c74fdd7c..74f43dae 100644
--- a/src/renderer/components/ft-input-tags/ft-input-tags.js
+++ b/src/renderer/components/ft-input-tags/ft-input-tags.js
@@ -55,9 +55,6 @@ export default Vue.extend({
         newList.splice(index, 1)
         this.$emit('change', newList)
       }
-    },
-    focusInput: function(e) {
-      document.getElementById(this.$refs.childinput.id).focus()
     }
   }
 })
diff --git a/src/renderer/components/ft-input-tags/ft-input-tags.vue b/src/renderer/components/ft-input-tags/ft-input-tags.vue
index 6bbc81ee..1ccdd6f3 100644
--- a/src/renderer/components/ft-input-tags/ft-input-tags.vue
+++ b/src/renderer/components/ft-input-tags/ft-input-tags.vue
@@ -2,23 +2,12 @@
   <div
     class="ft-input-tags-component"
   >
-    <div
-      class="input-tag-label-area"
-    >
-      <label
-        class="ft-input-tags-label"
-        @click="focusInput"
-      >{{ label }}</label>
-      <ft-tooltip
-        v-if="tooltip !== ''"
-        class="inputTagsTooltip"
-        :tooltip="tooltip"
-      />
-    </div>
-
     <ft-input
       ref="childinput"
       :placeholder="placeholder"
+      :label="label"
+      :show-label="true"
+      :tooltip="tooltip"
       :show-action-button="showActionButton"
       :select-on-focus="true"
       :force-action-button-icon-name="['fas', 'arrow-right']"
diff --git a/src/renderer/components/ft-input/ft-input.js b/src/renderer/components/ft-input/ft-input.js
index 5461fb05..40017f49 100644
--- a/src/renderer/components/ft-input/ft-input.js
+++ b/src/renderer/components/ft-input/ft-input.js
@@ -12,6 +12,10 @@ export default Vue.extend({
       type: String,
       required: true
     },
+    label: {
+      type: String,
+      required: null
+    },
     value: {
       type: String,
       default: ''
diff --git a/src/renderer/components/ft-input/ft-input.vue b/src/renderer/components/ft-input/ft-input.vue
index 727ffeeb..0757a09a 100644
--- a/src/renderer/components/ft-input/ft-input.vue
+++ b/src/renderer/components/ft-input/ft-input.vue
@@ -14,7 +14,7 @@
       v-if="showLabel"
       :for="id"
     >
-      {{ placeholder }}
+      {{ label || placeholder }}
       <ft-tooltip
         v-if="tooltip !== ''"
         class="selectTooltip"
-- 
2.37.1 (Apple Git-137.1)

```
</details>

PikachuEXE commented:
Found an issue
![image](https://user-images.githubusercontent.com/1018543/202352051-c2adb967-0c93-4618-8665-8ddc58c4becf.png)

petaded commented:
Have fixed the input button issue.
Also discovered an issue of my own ( when screen is below 680 px wide ) the tooltip positioning is ruined.
![image](https://user-images.githubusercontent.com/74797538/202448214-f4a0e009-545b-411e-bdc3-7ccc5f5d95d9.png)

Have fixed both problems,
Button position corrected:
![image](https://user-images.githubusercontent.com/74797538/202448306-1941743f-4afc-41f3-9aba-fd4fd3e09017.png)

Tooltip position corrected when on a small screen:
![image](https://user-images.githubusercontent.com/74797538/202448423-e6ee45d3-f88d-4366-9a94-848a392f0fba.png)

PikachuEXE commented:
@petaded I edited your comment to make the video playable inline
The trick is to have a blank line **before** the video URL

PikachuEXE reviewed (commented):
Issued fixed, accessibility update tested

absidue commented on src/renderer/components/ft-input-tags/ft-input-tags.vue:
```diff
@@ -0,0 +1,39 @@
+            class="removeTagButton"
+            tabindex="0"
+            role="button"
+            @click="removeTag"
+            @keydown.enter.prevent="removeTag"
```
```suggestion
            @click="removeTag(tag)"
            @keydown.enter.prevent="removeTag(tag)"
```

absidue commented on src/renderer/components/ft-input-tags/ft-input-tags.js:
```diff
@@ -0,0 +1,60 @@
+      // clear input box
+      this.$refs.childinput.handleClearTextClick()
+    },
+    removeTag: function (e) {
+      // Remove tag from list
+      let parent = e.srcElement.parentElement
+      if (parent.nodeName === 'svg') {
+        // for when user clicked the "path" rather than "svg"
+        parent = parent.parentElement
+      }
+      const tagName = parent.firstChild.outerText.trim()
```
```suggestion
    removeTag: function (tagName) {
```

petaded replied:
Hi @absidue thanks for taking a look, Your idea does indeed work and simplifies things. have pushed a commit including it.

absidue reviewed (commented):
Passing the tag through, means that you avoid having to read the DOM, haven't tested my code suggestion though, so please test it yourself.

petaded commented:
Thanks for info about the video @PikachuEXE I knew it was possible to show inline but didn't understand why it wasn't working.

Video works for me on brave, but not on firefox... odd. It should be h264 encoded.
```
> ffprobe accessiblity_test.mp4
.....
Stream #0:0[0x1](und): Video: h264 (High 4:4:4 Predictive) (avc1 / 0x31637661), yuv444p(progressive), 1520x652, 373 kb/s, 60 fps, 60 tbr, 16k tbn (default)
```

absidue commented on src/renderer/components/ft-list-lazy-wrapper/ft-list-lazy-wrapper.js:
```diff
@@ -36,11 +36,46 @@ export default Vue.extend({
+          }
+        }
+      }
+      return true
```
the previous check hid the item if `data` is undefined, this implementation doesn't

miangraham replied:
I believe the conflict resolutions in the rebase below will resolve this as well. See [L59](https://github.com/FreeTubeApp/FreeTube/commit/4d618e3b5459d72eafb9a98fca5c4a9b7a3d1182#diff-73b4b93ba9f89c56a9a2517a9ceb144b6fb10a00bdecb8c70b1291e4a77827b0R59).

petaded replied:
Yes this is fixed with the merge with latest development

miangraham commented:
@petaded I ran a rebase for this on development after my conflicting changes were merged. Tested that my option and yours still work afterward. Result over here, feel free to grab/squash:

https://github.com/miangraham/FreeTube/tree/add_option_to_hide_channels

petaded commented:
Hi all,

have merged latest development in including @miangraham changes. 
The merge has fixed the issue @absidue points out and the conflicts.

Would have used your provided merge @miangraham but i'd already done and pushed my merge when i saw your comment. thanks for the help though.

miangraham commented:
Ran a quick test with `npm run dev` post-merge and theirs/ours sides both work fine. Channels hidden by name, premieres hidden by flag. LGTM!

ChunkyProgrammer reviewed (approved):
Lgtm

BarbzYHOOL commented:
it has been released?

absidue commented:
This was merged on the 17th of December, the latest release 0.18.0 was on the 2nd of November...
